contracts/adapters/melonProtocol/MelonTokenAdapter.sol

```diff
@@ -0,0 +1,112 @@
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+pragma solidity 0.6.5;
+pragma experimental ABIEncoderV2;
+
+import { ERC20 } from "../../ERC20.sol";
+import { TokenMetadata, Component } from "../../Structs.sol";
+import { TokenAdapter } from "../TokenAdapter.sol";
+
+
+/**
+ * @dev Shares contract interface.
+ * Only the functions required for MelonAdapter contract are added.
+ * The Shares contract is available here
+ * github.com/melonproject/protocol/blob/master/contracts/fund/shares/Shares.sol.
+ */
+interface Shares {
+    function calcGav() external view returns (uint256);
+}
+
+
+/**
+ * @dev Registry contract interface.
+ * Only the functions required for MelonAdapter contract are added.
+ * The Registrycontract is available here
+ * @github.com/melonproject/protocol/blob/master/src/version/Registry.sol.
+ */
+interface Registry {
+    function getSymbol(address _asset) external view returns (string memory);
+    function getName(address _asset) external view returns (string memory);
+    function isFund(address _who) external view returns (bool);
+}
+
+/**
+ * @dev Accounting contract interface.
+ * Only the functions required for MelonAdapter contract are added.
+ * The Accounting contract is available here
+ * github.com/melonproject/protocol/blob/master/src/fund/accounting/Accounting.sol.
+ */
+interface Accounting{
+    function getFundHoldings() external view returns (uint[] memory, address[] memory);
+    function calcAssetGAV(address) external view returns (uint);
+}
+
+/**
+ * @title Token adapter for Melon Protocol by @codingsh.
+ * @dev Implementation of TokenAdapter interface.
+ * @author codingsh <codingsh@pm.me>
+ */
+contract MelonTokenAdapter is TokenAdapter {
+
+    address internal constant REGISTRY = 0xb9Cb55C9366a224647B7ff66252b3613185DA0B9;
+
+    /**
+     * @return TokenMetadata struct with ERC20-style token info.
+     * @dev Implementation of TokenAdapter interface function.
+     */
+    function getMetadata(address token) external view override returns (TokenMetadata memory) {
+        return TokenMetadata({
+            token: token,
+            name: Registry(REGISTRY).getName(token),
+            symbol: ERC20(token).symbol(),
+            decimals: ERC20(token).decimals()
+        });
+    }
+
+    /**
+     * @return Array of Component structs with underlying tokens rates for the given token.
+     * @dev Implementation of TokenAdapter interface function.
+     */
+    function getComponents(address token) external view override returns (Component[] memory) {  
+        uint[] memory quantities;
+        address[] memory fundHoldings;
+        try Accounting(token).getFundHoldings() returns (uint[] memory balance, address[] memory result) {
```

Comment:
`token` is not accounting address
accounting address should be retrieved with the hub contract

Reply:
done